chapters/refereecommands.adoc

```diff
@@ -17,8 +17,29 @@ When the halt command is issued, no robot is allowed to move or <<Ball Manipulat
 There is a grace period of 2 seconds for the robots to brake.
 
 .Usage
-The halt command allows the referee to interrupt the game immediately whenever an emergency occurs (for example when a robot gets out of contol). It is
-also used to recalibrate the vision software during a game if the vision expert considers it necessary and the referee agrees and for <<Robot Substitution, robot substitution>>. Additionally, the referee is free to issue the halt command at will.
+The halt command allows the referee to interrupt the game immediately whenever an emergency occurs (for example when a robot gets out of control). It is also used for <<Robot Substitution, robot substitution>>. Additionally, the referee is free to issue the halt command at will.
+
+Regarding the vision system, the game will be immediately halted in case of catastrophic failures, which includes:
+
+. Camera failures;
+. SSL-Vision crash;
+. Network failures;
+. Anything that causes a significant drop in the frame rate of the data;
+
+The game is continued in case of:
+
+. The ball is detected at least every 0.2 seconds while not being covered by some robot or shadow;
+. All robots on the field are detected at least every 0.5 seconds;
```

Comment:
I just ran some math on these numbers and I'd call them "way too high":

So, a robot can be misdetected for half a second during "normal game". Robots regularly drive with up to 4 m/s, giving 2 meters of "teleportation" in any tracking system. This might be acceptable for your own robots, but seems very bad for opponent robots. If you want to maximize the impact, let's assume the robots starts at 4 m/s and start to use it's acceleration of 4 m/s*s to get somewhere else during this half a second. They can end up at surprising positions somewhere on the field, that noone is prepared to not suddenly collide with them. In order to enter this "zone" where the robot could be, you have to slow down to be slower than this robot that's hiding, slowing down to at most 2 m/s.

While a lot of teams published some ideas about path finding, I haven't seen any team (yet) publish anything regarding "slow down obstacles" - that you could enter, but only with half of your normal speed. So most likely you'd have to build an obstacle for an invisible robot with diameter of more than 1 meter. That's **very large**, compared to 0.2 meters diameter of a robot. And the auto-ref might or might not be able to detect the collision, as well.

The same math for the ball: the ball accelerates almost instantly to 6.5 m/s, traveling approx. 1 meter while invisible. This is almost the whole distance between the defense area and the goal. Combine it with a robot that is invisible itself, and you have a very dangerous ball out of nowhere.

I'd suggest to tweak the numbers to 0.2 for the robots and 0.1 for the ball. Even half a meter is bad for the ball, but we need to put a point somwhere. 0.2 seconds for the robot makes sure that a standing robot will only move it's radius during a blackout, which sounds a lot more reasonable.

A further point: I just did some maths. I haven't checked what quality of raw data we have been using in 2019 or 2018, mostly because we don't use raw vision data very often. It might be worth to check if these numbers have been archived on previous competitions.

Reply:
I run some analysis's on past log files. The majority of data loss is indeed in the area of 100-200ms.

However, I'm not sure if those numbers are really what we want. It also depends on:

* the frequency of the data loss
* the speed of the object
* the location of the object 

If we loose several frames while an object is moving fast, this has more impact than when it is lying still. When a robot is close to the field border, the quality can be a bit worse in general, but robots can also not move fast in this area anyway.

In the mean time, I started creating a quality inspector that tries to detect data loss. Also, the TIGERs autoRef has a built-in quality inspector already.

I propose to change the whole rule to something like: 

Vision quality issues may only be reported to the vision expert by a mutually agreed on automatic software running on the vision expert PC, provided by the competition organizers or a team. Only the vision expert may ask the referee to stop the game in case of issues with the vision.
Teams may only complain on catastrophic failures of vision or network.

Optionally, we could add sth like:

If the game was halted due to reported vision or network failures and the failure could not be confirmed, the team has to take a timeout immediately.

Reply:
As the vision is a shared system among all teams they all share the drawbacks of a bad vision. If we assume vision problems to be normally distributed and field sides are swapped at half time each team has equally good or bad vision. Hence, I think the exact numbers for quality are of minor importance.
Nevertheless, I would recommend to change the metric to something like "number of detections within a second" or so. This still allows relatively long outtages in the worst case. I think that is okay and would rather encourage a "deal with it" mentality than babysitting every team for perfect vision. Missed detections are part of the game challenge and should be properly handled and filtered by the teams.

Reply:
@andre-ryll

While it is true that the shared vision system's errors should be distributed normally, that's not the whole point. On a normal Robocup, the final has much better vision than the first (friendly / group stage) games. That is - in part - due to the fact that teams report problems in the current calibration and they get fixed over time. While tracking and other tools can and will improve your software, worse inputs will always result in worse outputs and we should work towards a good detection quality during the competition.

The exact numbers for quality are and interesting side-advantage on this PR: Teams are able to test their software against the "worst possible" calibration and can try to work on this scenario. If these numbers are chosen correctly, they can hit the fine balance between "deal with it" and "impossible" as I stated in the first comment. I agree that (some) missed detections are part of the game, shadowing is aswell. I want to implement a "bad vision" simulator for our framework to be able to see the effect of repeatedly coming close to that threshold on our tracking - which as you might know already compensates for subperfect vision a lot.

These exact number also ensure that the vision will be improved during the matches, without the interruptions from buggy team software (sorry, we're at fault here :( ). Which I think is a good compromise.

Why do you think that Frames per second are the better metric that "time of gaps" - I think ultimately we would need both, because there's a big difference between one gap of 150 ms and only 6 detections per second, but I think that time between detections is even more important - as you might miss the robots acceleration for at least this "gap time". If it exceeds a threshold you're simply not able to avoid collisions. Do you have a number of frames per second in mind that seems appropriate?